### PR TITLE
Implement report table caching layer

### LINKED ISSUE #929: Implement report table caching layer
The table should contain the following columns
Report
- request id (pk, fk Request) int64
- validator (pk,fk Validator) string
- tx hash (fk Transaction) []byte
- reporter string

ReportDetail
- request id (pk, fk Report) int64
- validator (pk,fk Report) string
- External id (pk) int64
- DataSource id int64
- Data []byte
- Exitcode uint8

### PR DESCRIPTION
**Pull request by tansawit**
Fixes #929

### CODE REVIEW

File: chain/go.sum [hunk 1]
```diff
@@ -68,6 +68,7 @@ github.com/coreos/go-systemd v0.0.0-20190321100706-95778dfbb74e/go.mod h1:F5haX7
 github.com/coreos/pkg v0.0.0-20180928190104-399ea9e2e55f/go.mod h1:E3G3o1h8I7cfcXa63jLwjI0eiQQMgzzUDFVpN/nH/eA=
 github.com/cosmos/cosmos-sdk v0.37.8 h1:5KL3Cln5ezMKI+EvH7Dc9GgKgisdRzwbyrqiGPexkPc=
 github.com/cosmos/cosmos-sdk v0.37.8/go.mod h1:63Y8V75rRDzUrtRUwy+5cBOy5mr6xytI11PBk53tBrE=
+github.com/cosmos/cosmos-sdk v0.38.1 h1:DTuxIJeMpB//ydq+ObAjQgsaiwYBZ8T7NDzXjyiL1Kg=
```
Comment (sorawit) [inline chain/go.sum]: This has no effect on code correctness, but shouldn't be in this change.

File: chain/db/table.go [hunk 2]
```diff
@@ -57,3 +57,19 @@ type Transaction struct {
...
+type Report struct {
+	RequestID int64  `gorm:"primary_key;foreignkey:Request"`
+	Validator string `gorm:"primary_key;foreignkey:Validator"`
+	TxHash    []byte `gorm:"foreignkey:Transaction"`
+	Reporter  string `gorm:"not null"`
+}
+
+type ReportDetail struct {
+	RequestID    int64  `gorm:"primary_key;foreignkey:Report"`
+	Validator    string `gorm:"primary_key;foreignkey:Report"`
```
Comment (sorawit) [inline chain/db/table.go]: Do we need `foreignkey:Report`? I don't see it being one on other tables.

File: chain/db/reports.go [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+			return err
+		}
+	}
+
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
```
Comment (sorawit) [inline chain/db/reports.go]: Please always add trailing endlines.

File: chain/db/reports.go [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+			Data:         data.Data,
+			Exitcode:     data.ExitCode,
+		}).Error
+
+		if err != nil {
+			return err
+		}
+	}
+
+	if err != nil {
```
Comment (sorawit) [inline chain/db/reports.go]: This check should come right after the error is created. Move it to line 20.

**Review by sorawit — CHANGES_REQUESTED** [review]

File: chain/db/db.go [hunk 5]
```diff
@@ -50,6 +52,29 @@ func NewDB(dialect, path string, metadata map[string]string) (*BandDB, error) {
...
+
+	db.Model(&ReportDetail{}).AddForeignKey(
+		"request_id,validator",
+		"reports(request_id,validator)",
+		"RESTRICT",
+		"RESTRICT",
+	)
+
+	/*
+	db.Model(&ReportDetail{}).AddForeignKey(
```
Comment (sorawit) [inline chain/db/db.go]: Remove

**Review by sorawit — CHANGES_REQUESTED** [review]

**Review by sorawit — DISMISSED** [review]

File: chain/db/db.go [hunk 6]
```diff
@@ -51,13 +53,27 @@ func NewDB(dialect, path string, metadata map[string]string) (*BandDB, error) {
 		"RESTRICT",
 	)
 
+	db.Model(&Report{}).AddForeignKey(
+		"validator",
+    "validators(consensus_address)",
```
Comment (sorawit) [inline chain/db/db.go]: Fix code format

File: chain/db/db.go [hunk 7]
```diff
@@ -51,13 +53,27 @@ func NewDB(dialect, path string, metadata map[string]string) (*BandDB, error) {
...
+		"RESTRICT",
+		"RESTRICT",
+	)
+  
+  db.Model(&ReportDetail{}).AddForeignKey(
+		"request_id,validator",
+		"reports(request_id,validator)",
+    "RESTRICT",
```
Comment (sorawit) [inline chain/db/db.go]: also here

File: chain/db/db.go [hunk 8]
```diff
@@ -51,13 +53,27 @@ func NewDB(dialect, path string, metadata map[string]string) (*BandDB, error) {
...
+	)
+
 	db.Model(&Block{}).AddForeignKey(
 		"proposer",
 		"validators(consensus_address)",
 		"RESTRICT",
 		"RESTRICT",
 	)
-
+
```
Comment (sorawit) [inline chain/db/db.go]: also here

File: chain/db/db.go [hunk 9]
```diff
@@ -51,6 +53,20 @@ func NewDB(dialect, path string, metadata map[string]string) (*BandDB, error) {
 		"RESTRICT",
 	)
 
+	db.Model(&Report{}).AddForeignKey(
+		"validator",
+		"validators(consensus_address)",
+		"RESTRICT",
+		"RESTRICT",
+	)
+
```
Comment (taobun) [inline chain/db/db.go:69]: Add foreign key to the transaction by txhash

**Review by taobun — CHANGES_REQUESTED** [review]

**Review by sorawit — DISMISSED** [review]
Looks good.

File: chain/db/db.go [hunk 10]
```diff
@@ -51,6 +53,27 @@ func NewDB(dialect, path string, metadata map[string]string) (*BandDB, error) {
 		"RESTRICT",
 	)
 
+	db.Model(&Report{}).AddForeignKey(
+		"validator",
+		"validators(consensus_address)",
```
Comment (taobun) [inline chain/db/db.go]: operator_address

**Review by taobun — CHANGES_REQUESTED** [review]

**Review by taobun — APPROVED** [review]

**Review by sorawit — APPROVED** [review]

**Pull request merged**